### PR TITLE
Ensure that all uno controls defined dependency properties are using `FrameworkPropertyMetadata`

### LINKED ISSUE #5984: Ensure that all uno controls defined dependency properties are using `FrameworkPropertyMetadata`
Many of the controls defined inside of Uno need to have dependency properties are declared with `PropertyMetadata` instead of `FrameworkPropertyMetadata`. This generally causes issues like this https://github.com/unoplatform/uno/pull/5976.

## Note for implementers
- Changing this implies the creation of a runtime test or an analyzer that scans all dependency properties defined and checks if the metadata is of the right type
- Make changes to all properties metadata that are not of the right type

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #5984

## PR Type

- Bugfix
- Build or CI related changes

## What is the current behavior?

Usage of `FrameworkPropertyMetadata` isn't enforced over `PropertyMetadata` in CI, and some controls are not using `FrameworkPropertyMetadata`.

## What is the new behavior?

Enforce usage of `FrameworkPropertyMetadata` by introducing a new internal `Register` overload and ban the public overload.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
For now, build should fail since I didn't fix all usage of `PropertyMetadata`. Before I move forward with fixing all usages (which is a bit tedious), I want an initial review on the approach I've taken and its correctness.

**Review by dr1rrb — COMMENTED** [review]
I love the idea 😍, wondering though if we have an option to forcefully use a standard `PropertyMetadata` if needed?

**Comment by Youssef1313** [comment]
> I love the idea 😍, wondering though if we have an option to forcefully use a standard `PropertyMetadata` if needed?

The diagnostic can be suppressed when using `PropertyMetadata` is intentional. I expect this should be rare, right?

**Comment by Youssef1313** [comment]
@jeromelaban @dr1rrb I fixed the large portion of it. The remaining usages are in runtime tests and samples app. Should these remain the same and disable the diagnostic for them?

**Comment by davidjohnoliver** [comment]
> @jeromelaban @dr1rrb I fixed the large portion of it. The remaining usages are in runtime tests and samples app. Should these remain the same and disable the diagnostic for them?
> 
> There is also RegisterAttached, we should do the same for it too right? If so, do you prefer it in the same PR or separate PR?

For the runtime tests and SamplesApp, those ones are also compiled on Windows so they should (and must) remain the same.

**Comment by davidjohnoliver** [comment]
> Thanks @davidjohnoliver. Could someone look at the failure in CI? Is it a known flakiness or something related to my changes?

This one is not just run-of-the-mill flakiness:

**Uno.UI - CI (Unit Tests CSharp9)**
> RunMessage : The active test run was aborted. Reason: Test host process crashed : Process is terminated due to StackOverflowException.

Sounds related to this warning from the code analysis:

image

**Comment by Youssef1313** [comment]
@jeromelaban I think this is ready for review.

**Comment by MartinZikmund** [comment]
@Youssef1313 just a tip - I see you use "fix:" for commits fixing bugs in the PR scope, but it is better to just use "chore:" for these. "fix:" should be used for bug fixes for issues that are already in master. It is useful to avoid using "fix:" for these incremental adjustments, as we then use the prefixes to generate changelog, and this way the "fix" commits will show up under bug fixes.

**Comment by Youssef1313** [comment]
Thanks for the tip @MartinZikmund. I squashed and changed the commit message prefix to "chore:"

**Comment by MartinZikmund** [comment]
@Youssef1313 sorry, maybe I wrote it wrong - the initial commit would be "fix:" - the metadata is wrong in master and it fixes this. But the subsequent commits like fixing the StackOverflow would then be "chore" as they just fix an issue introduced by a previous commit in the PR 👍

**Comment by Youssef1313** [comment]
Thanks again for the clarification!

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationView.Properties.cs [hunk 1]
```diff
@@ -185,7 +185,7 @@ public DataTemplate HeaderTemplate
 		/// Identifies the HeaderTemplate dependency property.
 		/// </summary>
 		public static DependencyProperty HeaderTemplateProperty { get; } =
-			DependencyProperty.Register(nameof(HeaderTemplate), typeof(DataTemplate), typeof(NavigationView), new PropertyMetadata(null, OnPropertyChanged));
+			DependencyProperty.Register(nameof(HeaderTemplate), typeof(DataTemplate), typeof(NavigationView), new FrameworkPropertyMetadata(null, OnPropertyChanged));
```
Comment (davidjohnoliver) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationView.Properties.cs:188]: Be careful, DPs of type `DataTemplate` or `Style` should be marked `FrameworkPropertyMetadataOptions.ValueDoesNotInheritDataContext`.

Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationView.Properties.cs:188]: Thanks for noting, I'll fix that.

Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationView.Properties.cs]: Fixed. Note that some existing properties in master were already using `FrameworkPropertyMetadata` but not specifying the option. Example from master:

https://github.com/unoplatform/uno/blob/890207ec33affe032e5251ddc2276704520ba63b/src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.Properties.cs#L175-L180

Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationView.Properties.cs]: Also note that this PR only handles `DependencyProperty.Register`, but not `RegisterAttached`. If both should be handled, please let me know whether you want it in the same PR or in a separate PR. Thanks!

Comment (davidjohnoliver) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationView.Properties.cs]: I think it should be the same for `RegisterAttached()`, but it probably makes sense to put it in a separate PR and close this one off.

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Expander/Expander.Properties.cs [hunk 2]
```diff
@@ -45,7 +45,7 @@ public DataTemplate HeaderTemplate
 				nameof(HeaderTemplate),
 				typeof(DataTemplate),
 				typeof(Expander),
-				new PropertyMetadata(null));
+				new FrameworkPropertyMetadata(null, FrameworkPropertyMetadataOptions.ValueDoesNotInheritDataContext));
```
Comment (jeromelaban) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Expander/Expander.Properties.cs:48]: Maybe I missed some of the conversation, but why did you add this option?

Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/Expander/Expander.Properties.cs:48]: @jeromelaban This was requested in https://github.com/unoplatform/uno/pull/6630#discussion_r685865886

Comment (jeromelaban) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/Expander/Expander.Properties.cs:48]: Thanks!
